poisson2d/modern.f90

```diff
@@ -0,0 +1,83 @@
+module rho_interface
+  !! Define a scalar function over 2-space.
+  implicit none
+
+  private
+  public :: rho, dp
+
+  integer, parameter :: precision = 15, range = 307
+  integer, parameter :: dp = selected_real_kind(precision, range)
+
+  interface 
+    pure real(dp) module function rho(x,y)
+      !! Poisson equation inhomogeneous term.
+      implicit none
+      real(dp), intent(in) :: x,y
+    end function
+  end interface
+
+end module 
+
+submodule(rho_interface) rho_definition
+  implicit none
+contains
+  module procedure rho
+      !! To Do: give meaningful names to the magic numbers.
+      !! See https://en.wikipedia.org/wiki/Magic_number_(programming)#Unnamed_numerical_constants.
+      if (all([x,y]>0.6_dp .and. [x,y]<0.8_dp)) then
+        rho = 1._dp
+      else
+        rho = merge(-1._dp, 0._dp, all([x,y]>0.2_dp .and. [x,y]<0.4_dp))
+      end if
+  end procedure
+end submodule
+
+program poisson
+  !! Solve the 2D Poisson equation using a smoothing operation to iterate.
+  use rho_interface, only: rho, dp
+  implicit none
+
+  integer i,j
+  integer, parameter :: M=150
+  real(dp), parameter :: dx=0.01_dp
+  real(dp) :: t_start, rho_sampled(M,M)
+  
+  call cpu_time(t_start)
+
+  associate( dy => (dx) ) ! Associating with an expression provides immutable state so dy cannot be inadvertently redefined.
+
+    do concurrent(i=1:M, j=1:M)
+      rho_sampled(i,j) = rho(i*dx,j*dy)
+    end do
+ 
+    block ! Tighten the scoping to declutter the code above.
+      real(dp) :: delta_phi, t_end
+      real(dp), parameter :: epsilon0=8.85E-12_dp, tolerance=1E-6_dp
+      real(dp), dimension(M,M) :: phi_prime, phi
+      integer iteration
```

Comment:
Tiny style suggestion, for consistency with other declarations.

```suggestion
      integer :: iteration
```

Reply:
I usually only use the `::` where they are required, but I'm fine with adding it for consistency.